src/oas2/transformers/responses.ts

```diff
@@ -1,51 +1,72 @@
-import { IHttpOperationResponse } from '@stoplight/types';
+import { DeepPartial, Dictionary, IHttpOperationResponse, Optional } from '@stoplight/types';
 import { JSONSchema4 } from 'json-schema';
-import { chain, map, partial } from 'lodash';
-import { Response } from 'swagger-schema-official';
+import { chain, compact, map, partial } from 'lodash';
+import type { Response, Spec } from 'swagger-schema-official';
 
+import { isDictionary, maybeResolveLocalRef } from '../../utils';
+import { isResponseObject } from '../guards';
 import { getExamplesFromSchema } from './getExamplesFromSchema';
 import { translateToHeaderParams } from './params';
 
-function translateToResponse(produces: string[], response: Response, statusCode: string): IHttpOperationResponse {
-  const headers = translateToHeaderParams(response.headers || {});
-  const objectifiedExamples = chain(
-    response.examples || (response.schema ? getExamplesFromSchema(response.schema) : void 0),
-  )
-    .mapValues((value, key) => ({ key, value }))
-    .values()
-    .value();
-
-  const contents = produces.map(produceElement => ({
-    mediaType: produceElement,
-    schema: response.schema as JSONSchema4,
-    examples: objectifiedExamples.filter(example => example.key === produceElement),
-  }));
-
-  const translatedResponses = {
-    code: statusCode,
-    description: response.description,
-    headers,
-    contents,
-  };
+function responseTransformer(document: DeepPartial<Spec>) {
+  return function translateToResponse(
+    produces: string[],
+    response: unknown,
```

Comment:
Why is this now unknown if then you're casting it to `Response` on line 30?

Reply:
This was REALLY weird - if you pull this down you'll find that line 30 is the only place in this function that, for whatever reason, still thinks response is `unknown` even after the guard higher up in the function. All other use of `response` in this function correctly identifies that it's a `Response`. Any ideas?

Reply:
TypeScript is correct.

Although the arrow function has the context from outside, there's really no guarantee that the `isResponseObject` will be called *before* the arrow function. They are still separate entities. To fix this:

```ts
    const resolvedResponse = maybeResolveLocalRef(document, response);
    if (!isResponseObject(resolvedResponse)) return;

    const headers = translateToHeaderParams(resolvedResponse.headers || {});
    const objectifiedExamples = chain(
      resolvedResponse.examples || (resolvedResponse.schema ? getExamplesFromSchema(resolvedResponse.schema) : void 0),
    )
      .mapValues((value, key) => ({ key, value }))
      .values()
      .value();

    const contents = produces.map(produceElement => ({
      mediaType: produceElement,
      schema: resolvedResponse.schema as JSONSchema4,
      examples: objectifiedExamples.filter(example => example.key === produceElement),
    }));
```

```diff
- response = maybeResolveLocalRef(document, response);
+ const resolvedResponse = maybeResolveLocalRef(document, response);
```